_posts/2021-5-12-My-OTP-24-Highlights.md

```diff
@@ -0,0 +1,502 @@
+table
+3> ets:update_counter(table, k, 1).
+** exception error: bad argument
+     in function  ets:update_counter/3
+        called as ets:update_counter(table,k,1)
+        *** argument 2: not a key that exists in the table
+```
+
+That looks much better and now we can clearly see what the problem was!
+The standard logging formatters also include the additional information
+so that if this type of error happens in a production environment you will
+get the extra error information:
+
+```
+1> proc_lib:spawn(fun() -> ets:update_counter(table, k, 1) end).
+<0.94.0>
+=CRASH REPORT==== 10-May-2021::11:20:35.367023 ===
+  crasher:
+    initial call: erl_eval:'-expr/5-fun-3-'/0
+    pid: <0.94.0>
+    registered_name: []
+    exception error: bad argument
+      in function  ets:update_counter/3
+         called as ets:update_counter(table,k,1)
+         *** argument 1: the table identifier does
+                         not refer to an existing ETS table
+    ancestors: [<0.92.0>]
+```
+
+[EEP-54] is not only useful for error messages coming from BIFs but can be used
+by any application that wants to privide extra information about their exceptions.
+For example we have been working on providing better error information around
+`io:format` in [PR-4757](https://github.com/erlang/otp/pull/4757).
+
+[EEP-54]: https://www.erlang.org/erlang-enhancement-proposals/eep-0054.html
+[BIFs]: https://erlang.org/doc/reference_manual/functions.html#built-in-functions--bifs-
+
+# Improved receive optimizations #
+
+Since Erlang/OTP R14 (released in 2010) the Erlang compiler and run-time system
+have co-operated to specially optimize the pattern of code used by
+`gen_server:call` like functionality in order to not have to scan a potentially
+huge mailbox. The basic pattern looks like this:
+
+```erlang
+call(To, Msg) ->
+  Ref = make_ref(),
+  To ! {call, Ref, self(), Msg},
+  receive
+    {reply, Ref, Reply} -> Reply
+  end.
+```
+
+The compiler can from this figure out that when `Ref` is created, there can be
+no messages in the mailbox of the process that contain `Ref` and therefore it
+can skip all of those when receiving the `Reply`.
+
+This has always worked great in simple scenarios like this, but as soon as you
+have to make the scenarios a little more complex it tended to break the
+compilers analysis and you would end up scanning the entire mailbox. For example
+in the code below Erlang/OTP 23 will not optimize the receive.
+
+```erlang
+call(To, Msg, Async) ->
+  Ref = make_ref(),
+  To ! {call, Ref, self(), Msg},
+  if
+    Async ->
+      {ok, Ref};
+    not Async ->
+      receive
+        {reply, Ref, Reply} -> Reply
+      end
+  end.
+```
+
+That all changes with Erlang/OTP 24 though! Many more complex scenarios are now
+covered by the optimization and a new compiler flag has been added to tell the
+user if an optimization actually is done.
+
+```sh
+$ erlc +recv_opt_info test.erl
+test.erl:6: Warning: OPTIMIZED: reference used to mark
```

Comment:
I've only got a bad answer: we use the warning mechanism because it's the only way we can report these things at the moment.

I agree that it's rather confusing, and we'll look into fixing it in OTP 25.